### PR TITLE
Extend model regression tests for action prediction policies

### LINKED ISSUE #7835: Extend model regression tests for action prediction policies
Currently, the model regression tests can only run tests for the NLU components.
These should be extended to benchmark action prediction policies in Core as well.
We can use the datasets for Carbon bot, Sara, Financial demo as a starting point.
Note: This has to be done in collaboration with Infrastructure squad.
The current setup for model regression tests is described in this [notion document](https://www.notion.so/rasa/The-CI-for-model-regression-tests-92af7185e08e4fb2a0c764770a8e9095)

### PR DESCRIPTION
**Pull request by tczekajlo**
**Proposed changes**:
- Resolves #7835 

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by kedz** [comment]
Looks like all-core and possibly all-nlu are additional commands that you have implemented? If so we should have some explanation of them. Can you amend the explainer text for:

```
## Shortcut:
## You can use the "all" shortcut to include all available configurations or datasets
#
```

to be 

```
## Shortcuts:
## You can use the "all" shortcut to include all available configurations or datasets.
## You can use the "all-nlu" shortcut to include all available NLU configurations or datasets.
## You can use the "all-core" shortcut to include all available core configurations or datasets.
#
```

**Comment by kedz** [comment]
Is there a reason why action level F1 reports are showing no data while conversation level accuracy is 1.0 for 

- rule-augmemo
- rule-augmemo-TED
- rule-memo
- rule-memo-TED

policies?

**Comment by kedz** [comment]
Also, the insurance starter pack just came out so we should add 

`# - "insurance-demo" (NLU, Core)`

to the list of datasets.

**Comment by kedz** [comment]
Is it possible to make tensorflow run in deterministic mode like in the scheduled regression tests? To do this we would just need to set the environment variable `TF_DETERMINISTIC_OPS=1` [here](https://github.com/RasaHQ/rasa/blob/319eacec76fc6b8367d11ed41ae1c15d86a62d26/.github/workflows/ci-model-regression.yml#L299). Also the core policies will also be added to the scheduled regression tests right?

File: .github/workflows/ci-model-regression.yml [hunk 1]
```diff
@@ -269,15 +303,32 @@ jobs:
...
+            echo "::set-output name=train_run_time::$(gomplate -i '{{ $t := time.Parse time.RFC3339 (getenv "NOW_TRAIN") }}{{ (time.Since $t).Round (time.Second 1) }}')"
+
+            export NOW_TEST=$(gomplate -i '{{ (time.Now).Format time.RFC3339}}');
+            poetry run rasa test nlu --quiet -u ${DATASET_DIR}/$DATASET/${TEST_DIR} -m ${DATASET_DIR}/models/$DATASET/$CONFIG --out ${{ github.workspace }}/results/$DATASET/$CONFIG
+
+            echo "::set-output name=test_run_time::$(gomplate -i '{{ $t := time.Parse time.RFC3339 (getenv "NOW_TEST") }}{{ (time.Since $t).Round (time.Second 1) }}')"
+            echo "::set-output name=total_run_time::$(gomplate -i '{{ $t := time.Parse time.RFC3339 (getenv "NOW_TRAIN") }}{{ (time.Since $t).Round (time.Second 1) }}')"
+
+          elif [[ "${{ matrix.type }}" == "core" ]]; then
+            poetry run rasa train core --quiet -s ${DATASET_DIR}/$DATASET/$TRAIN_DIR/stories -c dataset/configs/$CONFIG -d ${DATASET_DIR}/${DATASET}/${DOMAIN_FILE}
```
Comment (kedz) [inline .github/workflows/ci-model-regression.yml:323]: I think `${DATASET_DIR}/$DATASET/$TRAIN_DIR/stories` could just be `${DATASET_DIR}/$DATASET/$TRAIN_DIR` right? That would still detect any story/rule files under the TRAIN_DIR directory I think. These files may not consistently be in a stories subdirectory.

File: .github/workflows/ci-model-regression.yml [hunk 2]
```diff
@@ -269,15 +303,32 @@ jobs:
...
+            echo "::set-output name=train_run_time::$(gomplate -i '{{ $t := time.Parse time.RFC3339 (getenv "NOW_TRAIN") }}{{ (time.Since $t).Round (time.Second 1) }}')"
 
-          export NOW_TEST=$(gomplate -i '{{ (time.Now).Format time.RFC3339}}');
-          poetry run rasa test nlu --quiet -u $DATASET/test -m models/$DATASET/$CONFIG --out ${{ github.workspace }}/results/$DATASET/$CONFIG
+            export NOW_TEST=$(gomplate -i '{{ (time.Now).Format time.RFC3339}}');
+            poetry run rasa test core -s ${DATASET_DIR}/$DATASET/$TRAIN_DIR/stories --out ${{ github.workspace }}/results/${{ matrix.dataset }}/$CONFIG
```
Comment (kedz) [inline .github/workflows/ci-model-regression.yml:327]: Same comment: 

I think `${DATASET_DIR}/$DATASET/$TRAIN_DIR/stories` could just be `${DATASET_DIR}/$DATASET/$TRAIN_DIR`.

File: .github/workflows/ci-model-regression.yml [hunk 3]
```diff
@@ -410,15 +495,32 @@ jobs:
...
+            echo "::set-output name=train_run_time::$(gomplate -i '{{ $t := time.Parse time.RFC3339 (getenv "NOW_TRAIN") }}{{ (time.Since $t).Round (time.Second 1) }}')"
+
+            export NOW_TEST=$(gomplate -i '{{ (time.Now).Format time.RFC3339}}');
+            poetry run rasa test nlu --quiet -u ${DATASET_DIR}/$DATASET/${TEST_DIR} -m ${DATASET_DIR}/models/$DATASET/$CONFIG --out ${{ github.workspace }}/results/$DATASET/$CONFIG
+
+            echo "::set-output name=test_run_time::$(gomplate -i '{{ $t := time.Parse time.RFC3339 (getenv "NOW_TEST") }}{{ (time.Since $t).Round (time.Second 1) }}')"
+            echo "::set-output name=total_run_time::$(gomplate -i '{{ $t := time.Parse time.RFC3339 (getenv "NOW_TRAIN") }}{{ (time.Since $t).Round (time.Second 1) }}')"
+
+          elif [[ "${{ matrix.type }}" == "core" ]]; then
+            poetry run rasa train core --quiet -s ${DATASET_DIR}/$DATASET/$TRAIN_DIR/stories -c dataset/configs/$CONFIG -d ${DATASET_DIR}/${DATASET}/${DOMAIN_FILE}
```
Comment (kedz) [inline .github/workflows/ci-model-regression.yml:515]: Same comment: 

I think `${DATASET_DIR}/$DATASET/$TRAIN_DIR/stories` could just be `${DATASET_DIR}/$DATASET/$TRAIN_DIR`.

File: .github/workflows/ci-model-regression.yml [hunk 4]
```diff
@@ -410,15 +495,32 @@ jobs:
...
+            echo "::set-output name=train_run_time::$(gomplate -i '{{ $t := time.Parse time.RFC3339 (getenv "NOW_TRAIN") }}{{ (time.Since $t).Round (time.Second 1) }}')"
 
-          export NOW_TEST=$(gomplate -i '{{ (time.Now).Format time.RFC3339}}');
-          poetry run rasa test nlu --quiet -u $DATASET/test -m models/$DATASET/$CONFIG --out ${{ github.workspace }}/results/$DATASET/$CONFIG
+            export NOW_TEST=$(gomplate -i '{{ (time.Now).Format time.RFC3339}}');
+            poetry run rasa test core -s ${DATASET_DIR}/$DATASET/$TRAIN_DIR/stories --out ${{ github.workspace }}/results/${{ matrix.dataset }}/$CONFIG
```
Comment (kedz) [inline .github/workflows/ci-model-regression.yml:519]: Same comment: 

I think `${DATASET_DIR}/$DATASET/$TRAIN_DIR/stories` could just be `${DATASET_DIR}/$DATASET/$TRAIN_DIR`.

Also I think this should be $TEST_DIR not $TRAIN_DIR

**Review by kedz — CHANGES_REQUESTED** [review]
Added several inline comments for minor changes.

**Comment by kedz** [comment]
Is it possible for me to run the regression tests now to test this out? I don't actually know how github actions work at all/how to specify that they should run off workflows defined in a branch and not main.

**Comment by tczekajlo** [comment]
> Is there a reason why action level F1 reports are showing no data while conversation level accuracy is 1.0 for
> 
> * rule-augmemo
> * rule-augmemo-TED
> * rule-memo
> * rule-memo-TED
> 
> policies? If they aren't showing up in the story_report.json I should look into why that is happening.

The reason why we can see `no data` is that it couldn't find the `micro avg` field in the `story_report.json`

**Comment by tczekajlo** [comment]
> Thanks @tczekajlo ! This looks pretty close! Is it possible for me to run the regression tests now to test this out? I don't actually know how github actions work at all/how to specify that they should run off workflows defined in a branch and not main.

You can run it within a PR, you can simply create a new PR that branch off `mrt_prediction_policies` or run some tests here.

File: .github/templates/model_regression_test_config_comment.tmpl [hunk 5]
```diff
@@ -9,6 +9,12 @@ Comment with a help message is added if a PR doesn't contain a comment
 with a configuration for the model regression tests.
 
  */ -}}
+{{ define "check_available_configuration" -}}
+NLU
+{{- if has .dataset "domain" -}}
```
Comment (tczekajlo) [inline .github/templates/model_regression_test_config_comment.tmpl:14]: @kedz Currently, I'm checking if the mapping file includes the `domain` field for a given dataset in order to determine if the dataset includes `core` or not

Is it correct? Or there is a better way to determine if a dataset includes `core` or not?

**Comment by tczekajlo** [comment]
@kedz Thanks a lot for your review, I've applied suggestions and it would be great if you can give it another look if you have a minute 😄

File: .github/templates/model_regression_test_config_comment.tmpl (hunk 5 above)
Comment (kedz) [reply .github/templates/model_regression_test_config_comment.tmpl:14]: I think that should work!

**Comment by kedz** [comment]
I am still confused why there is no action level data for 4 of the policies? I ran the tests locally (financial-demo with all of the core configs) and I am getting results for those fields. Additionally the Conversation level accuracies that I am getting are not 0.0 for the first two policies and not completely 1.0 for the others so I still think there is an issue with how the results are getting picked up.

**Comment by kedz** [comment]
Another issue I am realizing now is that we can specify the branch for the `training-data` repo but not for an external repository. For the time being I don't think its critical to be able to change the branch of the external repositories but we should instead just make the distinction explicitly stated so for an external repo like `financial-demo` it would say:
Dataset: `financial-demo`, Dataset repository branch: `main`, `training-data` repository branch: `regtest`

**Review by kedz — COMMENTED** [review]
Perhaps there is someone else on the engineering team that knows this templating language that could review the print_result_core pieces, because they don't appear to be printing correctly but I can't tell what the issue is.

**Comment by tczekajlo** [comment]
> I am still confused why there is no action level data for 4 of the policies? I ran the tests locally (financial-demo with all of the core configs) and I am getting results for those fields. Additionally the Conversation level accuracies that I am getting are not 0.0 for the first two policies and not completely 1.0 for the others so I still think there is an issue with how the results are getting picked up.

Let me take a closer look at it 👀

**Comment by tczekajlo** [comment]
> Commit: [a6c5ecf](https://github.com/RasaHQ/rasa/commit/a6c5ecf121bdb85e8f7f307882db36b023249bc2), [The full report is available as an artifact.](https://github.com/RasaHQ/rasa/actions/runs/770935181)
> 
> Dataset: `financial-demo`, Dataset repository branch: `regtest`
> 
> Dialog Policy Configuration	Action Level Micro Avg. F1	Conversation Level Accuracy	Run Time Train	Run Time Test
> `rule`	0.7218 (0.00)	0.5417 (0.00)	`15s`	`9s`
> `rule-TED`	0.7218 (0.00)	0.5417 (0.00)	`37s`	`30s`
> `rule-augmemo`	`no data`	1.0000 (0.00)	`14s`	`10s`
> `rule-augmemo-TED`	`no data`	1.0000 (0.00)	`38s`	`31s`
> `rule-memo`	0.9807 (0.00)	0.9167 (0.00)	`15s`	`11s`
> `rule-memo-TED`	0.9807 (0.00)	0.9167 (0.00)	`38s`	`31s`

@kedz Does it look better?

I've found a bug, and we were using a test directory as a training directory 🤦🏻

**Comment by kedz** [comment]
The results for this reg test (https://github.com/RasaHQ/rasa/pull/8338#issuecomment-825096504) are matching what I have locally so I think its in good shape! I started looking into why `rule-augmemo` and `rule-augmemo-TED` still have no data and for some reason the `story_report.json` doesn't have the "micro avg" field if the f-score is 1.0 and instead it creates a top level `accuracy` field instead. That doesn't really make any sense as to why it should behave the way but fixing it is a separate issue independent of the regression test (I'll make an issue about it). In the meantime, is it possible to use the top level accuracy (i.e. `cat story_report.json | jq .accuracy`) for the action result in the event that micro avg is not available?

**Comment by kedz** [comment]
/modeltest

dataset_branch: "regtest"
include:
 - dataset: ["all"]
   config: ["all-core"]

**Comment by tczekajlo** [comment]
/modeltest

```yml
include:
 - dataset: ["all"]
   config: ["all-core"]

```

**Comment by tczekajlo** [comment]
@kedz Is there anything to improve/change or all is good and the only thing we're waiting for is the `regtest` branch to be merged?

**Review by kedz — APPROVED** [review]
Looks good! Yeah I think we now just need to merge the regtest training-data branch. I'll follow up with @samsucik  about this now.

**Comment by kedz** [comment]
/modeltest

```yml
dataset_branch: "coreconfigs"
include:
 - dataset: ["all-core"]
   config: ["all-core"]

```

**Pull request merged**